`tail`: improve performance when stdin is piped

Joining7943 opened the pull request:
This pr intends to fix the performance issues described in #3842. In addition to the problems mentioned in #3842 with the `-n {+,-}{values}` option, the `-c {+,-}{values}` options were also very slow when reading in large files, which I also addressed in this pr. Basically, this solution uses fixed size byte reads processed in chunks instead of line based reads.

So far the performance is comparable to gnu's tail.

<details>
<summary>Latest benchmarks</summary>

```
❯ hyperfine --warmup 3 --output pipe -L prg tail,target/release/tail -L values 10,1000,100000,10000000 -L option c,n 'cat tests/fixtures/tail/random_ascii_505MB.BIG | {prg} -{option} +{values} -'
Benchmark 1: cat tests/fixtures/tail/random_ascii_505MB.BIG | tail -c +10 -
  Time (mean ± σ):     304.0 ms ±   3.3 ms    [User: 59.3 ms, System: 428.5 ms]
  Range (min … max):   299.6 ms … 310.8 ms    10 runs

Benchmark 2: cat tests/fixtures/tail/random_ascii_505MB.BIG | target/release/tail -c +10 -
  Time (mean ± σ):     318.1 ms ±   3.9 ms    [User: 73.2 ms, System: 430.4 ms]
  Range (min … max):   314.2 ms … 328.3 ms    10 runs

Benchmark 3: cat tests/fixtures/tail/random_ascii_505MB.BIG | tail -c +1000 -
  Time (mean ± σ):     307.4 ms ±   5.0 ms    [User: 57.7 ms, System: 432.5 ms]
  Range (min … max):   302.5 ms … 317.1 ms    10 runs

Benchmark 4: cat tests/fixtures/tail/random_ascii_505MB.BIG | target/release/tail -c +1000 -
  Time (mean ± σ):     317.0 ms ±   2.3 ms    [User: 76.2 ms, System: 425.5 ms]
  Range (min … max):   313.7 ms … 321.0 ms    10 runs

Benchmark 5: cat tests/fixtures/tail/random_ascii_505MB.BIG | tail -c +100000 -
  Time (mean ± σ):     310.2 ms ±  14.1 ms    [User: 60.6 ms, System: 427.9 ms]
  Range (min … max):   301.1 ms … 344.6 ms    10 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Benchmark 6: cat tests/fixtures/tail/random_ascii_505MB.BIG | target/release/tail -c +100000 -
  Time (mean ± σ):     316.0 ms ±   2.6 ms    [User: 61.9 ms, System: 438.7 ms]
  Range (min … max):   312.8 ms … 319.7 ms    10 runs

Benchmark 7: cat tests/fixtures/tail/random_ascii_505MB.BIG | tail -c +10000000 -
  Time (mean ± σ):     300.4 ms ±   1.2 ms    [User: 66.9 ms, System: 416.3 ms]
  Range (min … max):   298.5 ms … 302.3 ms    10 runs

Benchmark 8: cat tests/fixtures/tail/random_ascii_505MB.BIG | target/release/tail -c +10000000 -
  Time (mean ± σ):     316.4 ms ±   3.1 ms    [User: 65.6 ms, System: 434.3 ms]
  Range (min … max):   311.2 ms … 322.2 ms    10 runs

Benchmark 9: cat tests/fixtures/tail/random_ascii_505MB.BIG | tail -n +10 -
  Time (mean ± σ):     312.5 ms ±  17.0 ms    [User: 66.3 ms, System: 426.3 ms]
  Range (min … max):   301.8 ms … 352.7 ms    10 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Benchmark 10: cat tests/fixtures/tail/random_ascii_505MB.BIG | target/release/tail -n +10 -
  Time (mean ± σ):     317.7 ms ±   2.0 ms    [User: 62.7 ms, System: 440.2 ms]
  Range (min … max):   314.5 ms … 321.6 ms    10 runs

Benchmark 11: cat tests/fixtures/tail/random_ascii_505MB.BIG | tail -n +1000 -
  Time (mean ± σ):     306.8 ms ±   7.5 ms    [User: 56.3 ms, System: 436.3 ms]
  Range (min … max):   301.3 ms … 326.6 ms    10 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Benchmark 12: cat tests/fixtures/tail/random_ascii_505MB.BIG | target/release/tail -n +1000 -
  Time (mean ± σ):     320.9 ms ±  12.2 ms    [User: 70.9 ms, System: 431.6 ms]
  Range (min … max):   314.3 ms … 355.3 ms    10 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Benchmark 13: cat tests/fixtures/tail/random_ascii_505MB.BIG | tail -n +100000 -
  Time (mean ± σ):     300.4 ms ±   1.6 ms    [User: 60.8 ms, System: 422.5 ms]
  Range (min … max):   298.2 ms … 302.5 ms    10 runs

Benchmark 14: cat tests/fixtures/tail/random_ascii_505MB.BIG | target/release/tail -n +100000 -
  Time (mean ± σ):     313.9 ms ±   2.9 ms    [User: 71.0 ms, System: 427.7 ms]
  Range (min … max):   310.0 ms … 318.1 ms    10 runs

Benchmark 15: cat tests/fixtures/tail/random_ascii_505MB.BIG | tail -n +10000000 -
  Time (mean ± σ):     205.2 ms ±   2.3 ms    [User: 50.8 ms, System: 297.4 ms]
  Range (min … max):   201.8 ms … 209.8 ms    14 runs

Benchmark 16: cat tests/fixtures/tail/random_ascii_505MB.BIG | target/release/tail -n +10000000 -
  Time (mean ± σ):     203.3 ms ±   5.5 ms    [User: 77.9 ms, System: 291.6 ms]
  Range (min … max):   198.1 ms … 217.2 ms    14 runs

Summary
  'cat tests/fixtures/tail/random_ascii_505MB.BIG | target/release/tail -n +10000000 -' ran
    1.01 ± 0.03 times faster than 'cat tests/fixtures/tail/random_ascii_505MB.BIG | tail -n +10000000 -'
    1.48 ± 0.04 times faster than 'cat tests/fixtures/tail/random_ascii_505MB.BIG | tail -n +100000 -'
    1.48 ± 0.04 times faster than 'cat tests/fixtures/tail/random_ascii_505MB.BIG | tail -c +10000000 -'
    1.50 ± 0.04 times faster than 'cat tests/fixtures/tail/random_ascii_505MB.BIG | tail -c +10 -'
    1.51 ± 0.05 times faster than 'cat tests/fixtures/tail/random_ascii_505MB.BIG | tail -n +1000 -'
    1.51 ± 0.05 times faster than 'cat tests/fixtures/tail/random_ascii_505MB.BIG | tail -c +1000 -'
    1.53 ± 0.08 times faster than 'cat tests/fixtures/tail/random_ascii_505MB.BIG | tail -c +100000 -'
    1.54 ± 0.09 times faster than 'cat tests/fixtures/tail/random_ascii_505MB.BIG | tail -n +10 -'
    1.54 ± 0.04 times faster than 'cat tests/fixtures/tail/random_ascii_505MB.BIG | target/release/tail -n +100000 -'
    1.55 ± 0.04 times faster than 'cat tests/fixtures/tail/random_ascii_505MB.BIG | target/release/tail -c +100000 -'
    1.56 ± 0.04 times faster than 'cat tests/fixtures/tail/random_ascii_505MB.BIG | target/release/tail -c +10000000 -'
    1.56 ± 0.04 times faster than 'cat tests/fixtures/tail/random_ascii_505MB.BIG | target/release/tail -c +1000 -'
    1.56 ± 0.04 times faster than 'cat tests/fixtures/tail/random_ascii_505MB.BIG | target/release/tail -n +10 -'
    1.56 ± 0.05 times faster than 'cat tests/fixtures/tail/random_ascii_505MB.BIG | target/release/tail -c +10 -'
    1.58 ± 0.07 times faster than 'cat tests/fixtures/tail/random_ascii_505MB.BIG | target/release/tail -n +1000 -'

```

</details>

I already added some unit tests and integration tests to ensure basic correctness. However, this is currently still a draft and I'm writing on more tests and documenation.

Fixes #3842

tertsdiepraam commented:
Cool! It doesn't currently build on the MinRustV job, could you look into that? Also can we test this without those giant files with random noise? Maybe we can generate that data during the test?

tertsdiepraam commented on src/uu/tail/Cargo.toml:
```diff
@@ -18,12 +18,13 @@ path = "src/tail.rs"
-uucore = { version=">=0.0.15", package="uucore", path="../../uucore", features=["ringbuffer", "lines"] }
+memchr = "2.5.0"
+notify = { version = "=5.0.0-pre.16", features = ["macos_kqueue"] }
+uucore = { version = ">=0.0.15", package = "uucore", path = "../../uucore", features = ["ringbuffer", "lines"] }
```
I'm all for fixing this spacing, but preferably in another PR and across the entire project at the same time, because we have some scripts that assume a certain spacing (for upgrading the version number of uucore for instance)

Joining7943 replied:
fixed it. Should I create an issue for this?

tertsdiepraam reviewed (changes requested):
Whoops, selected the wrong option  :)

Joining7943 commented:
@tertsdiepraam Thanks for your feedback. I'll have a look into it as soon as possible.

tertsdiepraam commented on src/uu/tail/src/chunks.rs:
```diff
@@ -81,3 +92,424 @@ impl<'a> Iterator for ReverseChunks<'a> {
+    }
+
+    fn count_lines(&self) -> usize {
+        memchr::memchr_iter(self.delimiter, self.get_buffer()).count()
```
What is the performance improvement of this over just `self.get_buffer().iter().filter(|c| c == seld.delimiter).count()`?

Joining7943 replied:
Originally, I had the idea to search through the buffer with some fast c-like memory scan method. I know from the buffer, that it's a contiguous block and know the start and end location in the memory, so it looked appealing to me, but it seems to be unsafe in rust and I need more experience with rust to write unsafe code, I guess. I had no deeper look into memchr, but it was already in the secondary dependencies of uutils and after exchanging my slower for loop with it, I was happy and haven't asked any further. Maybe I missed something, however, I can try to exchange the dependency with your method, or you add it, if you like to.

Joining7943 replied:
I checked the difference, now, and counting with the suggested filter method has the same performance penalties like my for loop had. It's only noticable when the number of lines passed to `-n` is close or even to `10_000_000` in the benchmark below with a 505MB file.

<details>
<summary>benchmark of <code>iter().filter(|c| c == seld.delimiter).count()</code></summary>

```
❯ hyperfine --warmup 3 --output pipe -L prg tail,target/release/tail -L values 10,1000,100000,10000000 -L option c,n 'cat random_ascii_505MB.BIG | {prg} -{option} +{values} -'
Benchmark 1: cat random_ascii_505MB.BIG | tail -c +10 -
  Time (mean ± σ):     305.9 ms ±   4.8 ms    [User: 71.6 ms, System: 417.4 ms]
  Range (min … max):   302.5 ms … 318.8 ms    10 runs

Benchmark 2: cat random_ascii_505MB.BIG | target/release/tail -c +10 -
  Time (mean ± σ):     310.2 ms ±   1.8 ms    [User: 70.5 ms, System: 423.0 ms]
  Range (min … max):   306.2 ms … 312.6 ms    10 runs

Benchmark 3: cat random_ascii_505MB.BIG | tail -c +1000 -
  Time (mean ± σ):     310.2 ms ±  13.2 ms    [User: 66.7 ms, System: 421.8 ms]
  Range (min … max):   301.1 ms … 346.8 ms    10 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Benchmark 4: cat random_ascii_505MB.BIG | target/release/tail -c +1000 -
  Time (mean ± σ):     309.9 ms ±   2.5 ms    [User: 51.5 ms, System: 441.0 ms]
  Range (min … max):   306.3 ms … 314.9 ms    10 runs

Benchmark 5: cat random_ascii_505MB.BIG | tail -c +100000 -
  Time (mean ± σ):     305.1 ms ±   2.1 ms    [User: 67.8 ms, System: 420.4 ms]
  Range (min … max):   302.8 ms … 309.3 ms    10 runs

Benchmark 6: cat random_ascii_505MB.BIG | target/release/tail -c +100000 -
  Time (mean ± σ):     309.1 ms ±   2.2 ms    [User: 56.0 ms, System: 436.2 ms]
  Range (min … max):   305.8 ms … 312.5 ms    10 runs

Benchmark 7: cat random_ascii_505MB.BIG | tail -c +10000000 -
  Time (mean ± σ):     302.9 ms ±   1.5 ms    [User: 69.9 ms, System: 414.8 ms]
  Range (min … max):   299.7 ms … 305.8 ms    10 runs

Benchmark 8: cat random_ascii_505MB.BIG | target/release/tail -c +10000000 -
  Time (mean ± σ):     307.2 ms ±   1.7 ms    [User: 61.2 ms, System: 428.1 ms]
  Range (min … max):   304.6 ms … 310.9 ms    10 runs

Benchmark 9: cat random_ascii_505MB.BIG | tail -n +10 -
  Time (mean ± σ):     305.0 ms ±   2.7 ms    [User: 60.8 ms, System: 427.3 ms]
  Range (min … max):   301.5 ms … 309.5 ms    10 runs

Benchmark 10: cat random_ascii_505MB.BIG | target/release/tail -n +10 -
  Time (mean ± σ):     314.8 ms ±  10.4 ms    [User: 66.2 ms, System: 427.1 ms]
  Range (min … max):   305.9 ms … 339.7 ms    10 runs

Benchmark 11: cat random_ascii_505MB.BIG | tail -n +1000 -
  Time (mean ± σ):     306.2 ms ±   2.4 ms    [User: 73.9 ms, System: 415.8 ms]
  Range (min … max):   303.3 ms … 310.3 ms    10 runs

Benchmark 12: cat random_ascii_505MB.BIG | target/release/tail -n +1000 -
  Time (mean ± σ):     309.9 ms ±   2.9 ms    [User: 71.6 ms, System: 422.2 ms]
  Range (min … max):   308.2 ms … 318.0 ms    10 runs

Benchmark 13: cat random_ascii_505MB.BIG | tail -n +100000 -
  Time (mean ± σ):     305.3 ms ±   3.3 ms    [User: 76.0 ms, System: 412.5 ms]
  Range (min … max):   302.0 ms … 311.6 ms    10 runs

Benchmark 14: cat random_ascii_505MB.BIG | target/release/tail -n +100000 -
  Time (mean ± σ):     314.4 ms ±  15.0 ms    [User: 73.2 ms, System: 420.3 ms]
  Range (min … max):   307.1 ms … 356.7 ms    10 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Benchmark 15: cat random_ascii_505MB.BIG | tail -n +10000000 -
  Time (mean ± σ):     205.0 ms ±   2.6 ms    [User: 61.2 ms, System: 291.8 ms]
  Range (min … max):   199.7 ms … 207.4 ms    14 runs

Benchmark 16: cat random_ascii_505MB.BIG | target/release/tail -n +10000000 -
  Time (mean ± σ):     337.6 ms ±   7.8 ms    [User: 223.7 ms, System: 300.3 ms]
  Range (min … max):   329.7 ms … 353.1 ms    10 runs

Summary
  'cat random_ascii_505MB.BIG | tail -n +10000000 -' ran
    1.48 ± 0.02 times faster than 'cat random_ascii_505MB.BIG | tail -c +10000000 -'
    1.49 ± 0.02 times faster than 'cat random_ascii_505MB.BIG | tail -n +10 -'
    1.49 ± 0.02 times faster than 'cat random_ascii_505MB.BIG | tail -c +100000 -'
    1.49 ± 0.02 times faster than 'cat random_ascii_505MB.BIG | tail -n +100000 -'
    1.49 ± 0.03 times faster than 'cat random_ascii_505MB.BIG | tail -c +10 -'
    1.49 ± 0.02 times faster than 'cat random_ascii_505MB.BIG | tail -n +1000 -'
    1.50 ± 0.02 times faster than 'cat random_ascii_505MB.BIG | target/release/tail -c +10000000 -'
    1.51 ± 0.02 times faster than 'cat random_ascii_505MB.BIG | target/release/tail -c +100000 -'
    1.51 ± 0.02 times faster than 'cat random_ascii_505MB.BIG | target/release/tail -c +1000 -'
    1.51 ± 0.02 times faster than 'cat random_ascii_505MB.BIG | target/release/tail -n +1000 -'
    1.51 ± 0.02 times faster than 'cat random_ascii_505MB.BIG | target/release/tail -c +10 -'
    1.51 ± 0.07 times faster than 'cat random_ascii_505MB.BIG | tail -c +1000 -'
    1.53 ± 0.08 times faster than 'cat random_ascii_505MB.BIG | target/release/tail -n +100000 -'
    1.54 ± 0.05 times faster than 'cat random_ascii_505MB.BIG | target/release/tail -n +10 -'
    1.65 ± 0.04 times faster than 'cat random_ascii_505MB.BIG | target/release/tail -n +10000000 -'

```

</details>

tertsdiepraam replied:
Alright, thank you for investigating! `memchr` sounds like a good option then.

tertsdiepraam reviewed (commented):
Great work on the tests! I have commented 2 nits. It looks good to me, although I'm currently reading it all on my phone, so I can't really do a full review. It feels a bit verbose in places, but I think we can merge this and simplify it later if the need arises. Excellent work!

Joining7943 commented:
Thanks :)

Joining7943 commented on .gitignore:
```diff
@@ -15,3 +15,5 @@ lib*.a
 ### macOS ###
 .DS_Store
+
+**/*.BIG
```
Ok no problem, I can remove it.

Joining7943 commented:
I fixed the failing test on 32-bit systems by removing it. I don't think checking for memory limits for commandline arguments does make much sense and gnu's tail doesn't check either. 

I still don't know why all the windows tests are failing. All tests with piped in input fail as soon as some output on stdout is expected and it seems on windows systems tail produces no output, at all. I don't have a windows system around right now. Maybe you have an idea?

Joining7943 commented:
[83930ae](https://github.com/uutils/coreutils/pull/3874/commits/83930ae36229fffa3ee9b8b929ebe8ad6e2bb883) will fix the errors on windows. It won't fix #3845 and I don't know how this fix interferes with that pr.

Joining7943 commented:
> It feels a bit verbose in places, but I think we can merge this and simplify it later if the need arises.

@tertsdiepraam I'm currently doing some final refactorings, so if there is something you think needs a change, please you can tell me.

Joining7943 commented:
Thanks for your patience with this pr. I'm a little bit struggling with the last test result from github. I'm unsure if the failing tests in the cicd run https://github.com/uutils/coreutils/runs/8052898635?check_suite_focus=true for macos due to broken pipe is an error on my side. The exact error message is

```
2_Buildnightly (macos-latest, feat_os_macos).txt-2022-08-27T20:38:23.8022980Z thread 'test_tail::pipe_tests::test_pipe_when_negative_lines_option_given_no_newline_at_eof' panicked at 'failed to write to stdin of child: Broken pipe (os error 32)', tests/common/util.rs:1115:21
```

caused by a test using

```rust
new_ucmd!()
   .args(&["-n", "-0"])
   .pipe_in(random_string)
   .succeeds()
   .stdout_only("");
```

and because of the option `-n -0` producing no output, what eventually leads to the broken pipe. Shouldn't `tests/common/utils.rs` handle the broken pipe, or at least not fail the test because of it? Can you help, please?

The test for FreeBSD is failing too, because of

```
2022-08-27T20:59:58.0910250Z rsync warning: some files vanished before they could be transferred (code 24) at /AppleInternal/Library/BuildRoots/20d6c351-ee94-11ec-bcaf-7247572f23b4/Library/Caches/com.apple.xbs/Sources/rsync/rsync/main.c(1404) [generator=2.6.9]
2022-08-27T20:59:58.0943590Z ##[error]The process '/bin/bash' failed with exit code 24
```

I'm pretty sure there's something going wrong during the test setup, but I don't know what I can do about it.

tertsdiepraam commented:
We've indeed seen the broken pipe error before. It's something we'll have to investigate.

Joining7943 commented:
> We've indeed seen the broken pipe error before. It's something we'll have to investigate.

Ok. I'll leave the tests as they are, then.

> Whoops, selected the wrong option :)

I see you requested changes some days ago, but I haven't found out what it is. Is everything alright, now?

tertsdiepraam commented:
It's ok now. The changes were about fixing the build. I'll be able to do a full review in a few days.

Joining7943 commented:
ok. Thanks

tertsdiepraam commented on src/uu/tail/src/chunks.rs:
```diff
@@ -1,14 +1,24 @@
+//! Use [`ReverseChunks::new`] to create a new iterator over chunks of bytes from the file.
+
+// spell-checker:ignore (ToDO) filehandle
+use std::collections::vec_deque::{Iter, VecDeque};
```
`VecDeque` is also exported by `std::collections`:

```rust
use std::collections::VecDeque;
```

tertsdiepraam commented on src/uu/tail/src/chunks.rs:
```diff
@@ -81,3 +91,630 @@ impl<'a> Iterator for ReverseChunks<'a> {
+    ///
+    /// * `filehandle`: A [`BufReader`], where the inner must implement [`Read`]
+    ///
+    /// returns: `Result<Option<usize>, Box<dyn UError, Global>>`
```
The documentation for the argument and return type do not add much over the type signature here. I think we can remove them. The same goes for some other functions.

tertsdiepraam commented on src/uu/tail/src/chunks.rs:
```diff
@@ -81,3 +91,630 @@ impl<'a> Iterator for ReverseChunks<'a> {
+    pub(crate) fn fill(
+        &mut self,
+        reader: &mut BufReader<impl Read>,
+    ) -> UResult<Iter<Box<BytesChunk>>> {
```
I wonder if this might be cleaner if it doesn't return the iterator right away? I.e. this would just return `UResult<()>` and then you can just get the `VecDeque` via `self.chucks` after and iterate over it.

tertsdiepraam replied:
Or you could have the logic for printing all chucks that's now in `tail.rs` in a method and then you don't need to return the chunks at all.

Joining7943 replied:
I like the idea of having a printing function. Concerning the return type, I actually like having something useful returned if I can. The caller can still decide to ignore it. Would returning the total number of lines or total number of bytes contained in all chunks be an alternative for you?

tertsdiepraam replied:
I understand what you mean, but since were writing an internal function with a very well-defined goal, I think we should limit it to what we need and not add any unnecessary fluff. If the total number of lines is actually used somewhere then it's good, otherwise I'd prefer to leave it out.

tertsdiepraam commented on src/uu/tail/src/chunks.rs:
```diff
@@ -81,3 +91,630 @@ impl<'a> Iterator for ReverseChunks<'a> {
+        };
+
+        // calculate the number of lines to skip in the chunk
+        let skip_lines = if self.num_print >= self.lines {
+            0
+        } else {
+            (self.lines - self.num_print) as usize
+        };
```
You could look into `saturating_sub` for this, something like:
```rust
self.lines.saturating_sub(self.num_print) as usize
```

Joining7943 replied:
nice

tertsdiepraam commented on src/uu/tail/src/chunks.rs:
```diff
@@ -81,3 +91,630 @@ impl<'a> Iterator for ReverseChunks<'a> {
+}
+
+#[cfg(test)]
+mod tests {
```
I love the tests!

tertsdiepraam commented on src/uu/tail/src/chunks.rs:
```diff
@@ -81,3 +91,630 @@ impl<'a> Iterator for ReverseChunks<'a> {
+        let mut chunk = Box::new(LinesChunk::new(self.delimiter));
+
+        while (chunk.fill(reader)?).is_some() {
+            self.lines += chunk.lines as u64;
```
Maybe `self.lines` could be `usize`? It seems to be used with other `usize` values a lot.

Joining7943 replied:
you're right, the casting looks cumbersome but I wanted to support tailing files > 4GB on 32-bit systems, therefore the difference of LinesChunkBuffer.lines as u64 and LinesChunk.lines as usize. The u64 originates from FilterMode::Lines(count, ...) and FilterMode::Bytes(count), both `count` variables being u64. The usize is sufficient for LinesChunk.lines because the max value for lines is bytes, and bytes max value is BUFFER_SIZE which is a usize. Would it help updating the docs to include some infos about the choice of LinesChunkBuffer.lines as u64?

tertsdiepraam commented on src/uu/tail/src/chunks.rs:
```diff
@@ -1,14 +1,24 @@
-//! Iterating over a file by chunks, starting at the end of the file.
+//! Iterating over a file by chunks, either starting at the end of the file with [`ReverseChunks`]
```
This file should have the copyright info at the top:
```rust
//  * This file is part of the uutils coreutils package.
//  *
//  * For the full copyright and license information, please view the LICENSE
//  * file that was distributed with this source code.
```

tertsdiepraam commented on src/uu/tail/src/chunks.rs:
```diff
@@ -1,14 +1,24 @@
+/// buffer assume that the target system's usize is greater than this BUFFER_SIZE, and therefore
+/// convert from u64 to usize as long as it is known, that the value resides somewhere between 0
+/// and the BUFFER_SIZE.
+pub const BUFFER_SIZE: usize = 8192;
```
Could you explain in the doc how you got to this number?

tertsdiepraam commented on src/uu/tail/src/chunks.rs:
```diff
@@ -81,3 +91,630 @@ impl<'a> Iterator for ReverseChunks<'a> {
+    }
+}
+
+#[derive(Debug)]
```
Two very nitpicky comments, but I like the doc comment above the derives and `pub(crate)` and `pub` are used for similar things in this file. Generally, because `uutils` is not a library, simply `pub` is good enough in my opinion.

tertsdiepraam commented on tests/common/random.rs:
```diff
@@ -0,0 +1,331 @@
+//  * For the full copyright and license information, please view the LICENSE
+//  * file that was distributed with this source code.
+
+#![allow(dead_code)]
```
Do we still need this? Can we remove (or export) `random_char`?

Joining7943 replied:
I removed it.

Joining7943 replied:
I removed it.

tertsdiepraam reviewed (changes requested):
Alright, it looks excellent. All my comments are just small style things, nothing with the implementation. If you could fix those up I think we can merge this.

jhscheer commented on src/uu/tail/src/tail.rs:
```diff
@@ -30,11 +31,13 @@ use chunks::ReverseChunks;
 use notify::{RecommendedWatcher, RecursiveMode, Watcher, WatcherKind};
-use std::collections::{HashMap, VecDeque};
+use std::cmp::Ordering;
+use std::collections::vec_deque::VecDeque;
+use std::collections::HashMap;
```
```suggestion
use std::collections::{HashMap, VecDeque};
```

jhscheer replied:
```suggestion
use std::collections::{HashMap, VecDeque};
```

jhscheer commented on src/uu/tail/src/tail.rs:
```diff
@@ -30,11 +31,13 @@ use chunks::ReverseChunks;
-use std::fmt;
 use std::fs::{File, Metadata};
-use std::io::{stdin, stdout, BufRead, BufReader, Read, Seek, SeekFrom, Write};
+use std::io;
+use std::io::{stdin, stdout, BufRead, BufReader, BufWriter, Read, Seek, SeekFrom, Write};
```
```suggestion
use std::io::{self, stdin, stdout, BufRead, BufReader, BufWriter, Read, Seek, SeekFrom, Write};
```

jhscheer commented on tests/by-util/test_tail.rs:
```diff
@@ -2435,6 +2423,658 @@ fn test_illegal_seek() {
+    use super::*;
+    use crate::common::random::*;
+    use rand::distributions::Alphanumeric;
+    const CHUNK_BUFFER_SIZE: usize = 8192;
```
```suggestion
use tail::chunks::BLOCK_SIZE as CHUNK_BUFFER_SIZE;
```
together with `pub mod chunks;` in `tail.rs`

Joining7943 replied:
Besides using BUFFER_SIZE instead of BLOCK_SIZE, I applied your changes

jhscheer commented on tests/by-util/test_tail.rs:
```diff
@@ -2435,6 +2423,658 @@ fn test_illegal_seek() {
+            .succeeds()
+            .stdout_only(test_string);
+
+        #[cfg(all(not(target_os = "macos")))]
```
I tried this (and the other cfgs) on a macOS system. If uncommented the test sometimes succeeds and sometimes fails. If it fails it runs into the `!todo()` at `tail.rs:434:17`.
It's very strange, that it fails only sometimes. I suspect it has something to do with the handling of the legacy `+` syntax.

Joining7943 replied:
That's exaclty the same behaviour I observed when running the Github Actions CI jobs. The tests failed randomly on macos when no output was expected, at least this is what I suspected. Some tests also failed in cases of `-n -0` or `-c -0`.

jhscheer replied:
I did some more testing on macOS.
The failures will also sometimes (10-20% of the time) happen on the `main` branch.
However, they will only happen inside the test framework, i.e. this will never trigger a failure:
```bash
for i in {1..100}; do echo "a\nb\n" | target/debug/tail -n +3; done
```
but this will fail regularly:
```rust
let test_string = "a\nb\n";
new_ucmd!()
    .args(&["-n", "+3"])
    .pipe_in(test_string)
    .succeeds()
    .no_stdout()
    .no_stderr();
```

Joining7943 replied:
Strange indeed. So you think it has something to do with the test system? It's maybe better to move this to an own issue.

Joining7943 replied:
I've created an issue #3895 and added infos which might help narrowing down the error.

Joining7943 commented:
Thanks for your great review. I'll have a look at it asap.

Joining7943 commented:
I applied all suggested changes. There's maybe a little bit of a confusion around the BUFFER_SIZE. @jhscheer I added some explanatory comments in `chunks.rs` and I hope this will explain why I used 8192 and not BLOCK_SIZE.

tertsdiepraam commented on src/uu/tail/src/chunks.rs:
```diff
@@ -81,3 +96,604 @@ impl<'a> Iterator for ReverseChunks<'a> {
+    bytes: usize,
+}
+
+// Silence a clippy warning about a missing default trait implementation, because a user might
```
I think this can be solved with `derive(Default)`. And if not, we might be better off `allow`ing the warning on the struct.

Joining7943 replied:
I allowed the warning

tertsdiepraam reviewed (approved):
A few more suggestions based on the new changes, but I'll still approve because I think it's ready once those are fixed (mostly just the commented clippy attribute).

Joining7943 commented on tests/common/random.rs:
````diff
@@ -0,0 +1,317 @@
+/// println!("Random chars: {}", String::from_utf8(vec).unwrap());
+/// ```
+#[derive(Clone, Copy, Debug)]
+#[cfg_attr(feature = "serde1", derive(serde::Serialize, serde::Deserialize))]
````
It's not used anywhere. I've seen it in the rand crate when I started working on the random module. I didn't knew at this point that we don't need it in the end.

Joining7943 commented on tests/common/random.rs:
```diff
@@ -0,0 +1,317 @@
+//  * For the full copyright and license information, please view the LICENSE
+//  * file that was distributed with this source code.
+
+// #![allow(dead_code)]
```
Oops. Sorry, this happened by accident.

Joining7943 commented:
ok. I applied the latest chage requests.

Joining7943 commented:
Any idea why some tests related to uu_chgrp are failing all of the sudden?

sylvestre commented:
Dunno but I opened this bug:
https://github.com/uutils/coreutils/issues/3890

tertsdiepraam commented:
Looks like we're hitting a `todo!()` on mac/nightly. Could you check whether this has anything to do with your changes?

Joining7943 commented:
This doesn' have to do with my changes directly. In the `tail.rs` file I only changed code in unbounded_tail(). I guess it has something to do with the tests I added which test these commandline options more frequently. @jhscheer did some background checks and I added my findings to the issue #3895. See also here https://github.com/uutils/coreutils/pull/3874#discussion_r960938955. To summarize so far, it happens randomly and it happens on macos more frequently than on linux and it has something to do with the test system. I already switched off some of the tests on macos but the `-n +0` and `-c +0` options are also affected. I'll also switch off these tests for above command line options on macos and linux until it is fixed.

Joining7943 commented:
I decided to switch off the pipe tests on macos entirely, since also other values than `+0` to the `-c` or `-n` option are affected at random.

Joining7943 commented:
Beside trying to fix #3910, I would like to try something else and ignore the write errors to stdin. I readded macos to the pipe tests.

Joining7943 commented:
The tests look promising, no broken pipe errors anymore. The android test is failing but I don't see a relation to this pr. @tertsdiepraam Do you? I had concerns that the input isn't transferred through the pipe completely in case of a write error in `util`, but looks like this isn't the case. From my side this pr is safe to merge. Do you have any concerns?

tertsdiepraam commented:
The Android failure is probably this: https://github.com/uutils/coreutils/issues/3911. I'll do a final read-through of the PR.

tertsdiepraam reviewed (approved):
Looks all good now!

sylvestre commented:
arf :(
you didn't squash the 34 commits :(

Joining7943 commented:
@sylvestre sorry :( I didn't want to force push until everything's alright. I should have written something ...

sylvestre commented:
no worries, I am taking care of that

Joining7943 commented:
uff. ok

sylvestre commented:
Push here:
https://github.com/uutils/coreutils/commit/2658f8ae5badb3a5e55968ec396a29c891e0c795

What I had done:
$ git reset --hard b39f5239e766d1b105588fbeebf3412b1ddb04ea 
$ git push origin main -f 
(after giving me permissions)
$ wget https://patch-diff.githubusercontent.com/raw/uutils/coreutils/pull/3874.diff
$ git commit 
(and I just removed my permissions)

Joining7943 commented:
thanks and sorry again

sylvestre commented:
no worries :)

tertsdiepraam commented:
Whoops, sorry! @sylvestre

sylvestre commented:
no problem :)
in case you don't know, github allows you to do that directly in the interface :)

tertsdiepraam commented:
Yeah I just forgot to check the commits :)

Joining7943 commented:
@sylvestre yeah thanks but I don't trust such interfaces ;) I'm using git almost only on the command line.
